xio-core/src/test/java/com/xjeffrose/xio/http/ClientConnectionManagerIntegrationTest.java

```
@@ -0,0 +1,95 @@
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ClientConnectionManagerIntegrationTest extends Assert {
+
+  private class HollowChannelHandler extends SimpleChannelInboundHandler<String> {
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx, String msg) throws Exception {}
+  }
+
+  private ClientConnectionManager subject;
+  private MockWebServer server;
+
+  private ClientConnectionManager subjectFactory(boolean shouldSucceed) {
+    ClientChannelConfiguration channelConfiguration =
+        new ClientChannelConfiguration(new NioEventLoopGroup(), NioSocketChannel.class);
+
+    File configFile;
+    if (shouldSucceed) {
+      configFile = new File("src/test/resources/ClientConnectionManagerTestSucceed.conf");
+    } else {
+      configFile = new File("src/test/resources/ClientConnectionManagerTestFail.conf");
+    }
+    Config config = ConfigFactory.parseFile(configFile);
+    ClientConfig clientConfig = new ClientConfig(config);
+    ClientState clientState = new ClientState(channelConfiguration, clientConfig);
+
+    ClientChannelInitializer clientChannelInit =
+        new ClientChannelInitializer(clientState, () -> new HollowChannelHandler(), null);
+    return new ClientConnectionManager(clientState, clientChannelInit);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    if (server != null) {
+      server.close();
+    }
+  }
+
+  @Test
+  public void testInitialConditions() {
+    subject = subjectFactory(true);
+    assertEquals(subject.connectionState(), ClientConnectionState.NOT_CONNECTED);
+  }
+
+  @Test
+  public void testConnectingFailingConnection() throws Exception {
+    subject = subjectFactory(false);
+    // don't set up fake origin backend server so we can connect to it
+    Future<Void> connectionResult = subject.connect();
+    assertEquals(ClientConnectionState.CONNECTING, subject.connectionState());
+    connectionResult.awaitUninterruptibly(60000);
```

60 seconds is in the client configuration for conn timeout ... maybe lower it in the test config

---

yeah so, it takes up to that long for the connection to time out, we could configure the client pipeline to have a configurable timeout, but that should be done in another ticket since that is non-zero amount of work.

---

I will try an invalidly formed ip address to see if i can get the channel connect to fail faster

---

ok using a bogus ip address lets us fail the connection faster, will update pr

---

updated